poc/observe.py

```diff
@@ -0,0 +1,865 @@
+        return type(self) is type(other) and self.notify == other.notify
+
+    def event_factory(self, trait_list, index, removed, added):
+        return ListObserverEvent(trait_list, index, removed, added)
+
+    def iter_this_targets(self, object):
+        # object should be a TraitListObject
+        yield object
+
+    def iter_next_targets(self, object):
+        for item in object:
+            if is_notifiable(item):
+                yield item
+
+    def get_change_notifier(self, callback, path, target, dispatcher):
+        return ListenerChangeNotifier(
+            listener_callback=self.change_callback,
+            actual_callback=callback,
+            path=path,
+            target=target,
+            event_factory=self.event_factory,
+            dispatcher=dispatcher,
+        )
+
+    @staticmethod
+    def change_callback(event, callback, path, target, dispatcher):
+        logger.debug(
+            "Handling list change. "
+            "({!r})".format(
+                (event.new, event.removed, event.added, path.node))
+        )
+        for item in event.removed:
+            if is_notifiable(item):
+                remove_notifiers(
+                    object=item,
+                    callback=callback,
+                    path=path,
+                    target=target,
+                    dispatcher=dispatcher,
+                )
+        for item in event.added:
+            if is_notifiable(item):
+                add_notifiers(
+                    object=item,
+                    callback=callback,
+                    path=path,
+                    target=target,
+                    dispatcher=dispatcher,
+                )
+
+    def trait_added_matched(self, object, name, trait):
+        """ Return true if an added trait should be handled by this listener
+        """
+        return False
+
+
+class DictValueListener(BaseListener):
+    # TODO: Define this class taking into account
+    # what is going to happen in TraitDictObject
+
+    def __init__(self, notify):
+        self.notify = notify
+
+
+class ListenerPath:
+
+    def __init__(self, node, nexts=()):
+        self.node = node
+        self.nexts = nexts
+
+    def __eq__(self, other):
+        """ Return true if a given ListenerPath is equivalent to this one.
+        """
+
+        # FIXME: The following is a draft.
+        # We need to handle cycles!
+
+        if other is self:
+            return True
+        if type(other) is not type(self):
+            return False
+        if self.node != other.node:
+            return False
+        if self.nexts == other.nexts:
+            return True
+        if len(self.nexts) != len(other.nexts):
+            return False
+        for nexts in itertools.permutations(other.nexts, len(other.nexts)):
+            if all(n1 == n2 for n1, n2 in zip(self.nexts, nexts)):
+                return True
+        return False
+
+    @classmethod
+    def from_nodes(cls, node, *nodes):
```

Comment:
See #969 for constructing `ListenerPath` in a more user-friendly manner. This method might be useful to be written in the tests though.